### ISSUE TITLE
path type different than std::fs::File

### ISSUE THREAD

**Issue by dbregman** [andrewhickman/fs-err#34]
std::fs::File uses `AsRef<Path>` as the type for passing paths while fs_err uses `Into<PathBuf>`

This makes fs_err not be a "drop in replacement": When I pass an `AsRef<Path>` to fs_err::File::open, it does not compile

**Comment by LPGhatguy** [comment]
This is true. Maybe we should qualify this statement with an asterisk and explain how we differ?

In many cases, our API is still a drop-in replacement. By using `Into<PathBuf>` instead of `AsRef<Path>`, we're able to take ownership of the existing `PathBuf` in error cases, which is a handy optimization.

**Comment by kellpossible** [comment]
I also hit this, and had to rework some code. I'm guessing the error types need to own the path and that's why this is done?

**Comment by TonalidadeHidrica** [comment]
Is it really necessary that the error should own the path?  Can't it just store the reference to the path instead?

**Comment by hniksic** [comment]
@TonalidadeHidrica The error must own the path, it cannot be a reference because that would make both the wrapper and the error non-`'static`. Requiring `Into<PathBuf>` rather than `AsRef<Path>` is basically an optimization, allowing `fs_err` to avoid an allocation when you pass an actual `PathBuf`.

This issue should probably be resolved by a PR that improves the docs to explain this, and show how to change the code that doesn't compile due to the difference. (The change is trivial, but can be perplexing to a newbie.)

**Comment by andrewhickman** [comment]
Its worth noting that all uses of `Into<PathBuf>` are for constructing an `fs_err::File`, which owns a PathBuf even in the happy path. All the other functions were changed to `AsRef<Path>` and will allocate only in the failure path.

I'd be happy to accept a PR to improve docs around this
